File: src/147.insertion_sort_list.kt
import other.ListNode

fun main() {
    println(insertionSortList(ListNode(4).apply {
        next = ListNode(2).apply { next = ListNode(1).apply { next = ListNode(3) } }
    }))
    println(insertionSortList(ListNode(-1).apply {
        next = ListNode(5).apply {
            next = ListNode(3).apply { next = ListNode(4).apply { next = ListNode(0) } }
        }
    }))
}

fun insertionSortList(head: ListNode?): ListNode? {
    val newHead = ListNode(0)
    var currentNode = head
    var indexNode: ListNode?
    var temp: ListNode?

    while (currentNode != null) {
        indexNode = newHead
        while (indexNode?.next != null && indexNode.next!!.`val` < currentNode.`val`) {
            indexNode = indexNode.next
        }

        temp = currentNode.next
        currentNode.next = indexNode?.next
        indexNode?.next = currentNode
        currentNode = temp
    }

    return newHead.next
}

File: lib/147.insertion_sort_list.dart
import 'other/list_node.dart';

void main() {
  print(insertionSortList(
      ListNode(4)..next = (ListNode(2)..next = (ListNode(1)..next = ListNode(3)))));
  print(insertionSortList(ListNode(-1)
    ..next = (ListNode(5)..next = (ListNode(3)..next = (ListNode(4)..next = ListNode(0))))));
}

ListNode? insertionSortList(ListNode? head) {
  final newHead = ListNode(0);
  ListNode? indexNode;
  ListNode? tempNode;

  while (head != null) {
    indexNode = newHead;

    while (indexNode?.next != null && indexNode!.next!.val < head.val) {
      indexNode = indexNode.next;
    }

    tempNode = head.next;
    head.next = indexNode?.next;
    indexNode?.next = head;
    head = tempNode;
  }

  return newHead.next;
}
